README.md
[![mPlane](http://www.ict-mplane.eu/sites/default/files//public/mplane_final_256x_0.png)](http://www.ict-mplane.eu/)

#mPlane nodeJS reference library 
[![](https://travis-ci.org/finvernizzi/mplane.svg)](https://travis-ci.org/finvernizzi/mplane)


This is the [mPlane](http://www.ict-mplane.eu/) nodejs library. 
The architecture and software structure is freely inspired by [mPlane reference implementation](http://fp7mplane.github.io/protocol-ri/) written in python by Brian Trammell <brian@trammell.ch>.


#Installation

`npm install mplane`

#Usage example
In this example we set a capability for a probe. 

```javascript
var mplane = require('mplane');

var MY_IP="192.168.0.1";

// Initialize available primitives from a registry
// Can be any valid URL or local file
mplane.Element.initialize_registry("ict-mplane.eu/registry/demo");

// Create a new Capability object
var probeCapability = new mplane.Capability();
// Set time constraints
probeCapability.set_when("now ... future / 1s");

// Add parameters with constraints
probeCapability.add_parameter({
    type:"destination.ip4",
    constraints: "192.168.0.0/16"
}).add_parameter({
        type:"source.ip4",
        constraints:MY_IP
});

// Results
probeCapability.add_result_column("delay.twoway");

// Some metadata
probeCapability.set_metadata_value("System_type","Just a TEST")
    .set_metadata_value("System_version","0.1a")
    .set_metadata_value("System_ID","Pinger").update_token();
// A label to identify the capability
probeCapability.set_label("This_is_a_pinger");


```

#Interaction with the RI
The mPlane Reference Implementation, written in python, can be find [here](http://fp7mplane.github.io/protocol-ri/).
In order to correctly interact with any component using the RI messages formats, two method have been exposed in the library, to decode messages generated by the RI and to send messages to a RI component.

##from_dict
This is a global method exposed in the mplane library. It sould be used to import any raw message received from an mPlane RI component.
It will generate an mplane object of the type contained in the message.
In the following example the data received as a body of an HTTP communication is imported in an mPlane object.

```javascript
var result = mplane.from_dict(body);
```

## to_dict
This method, defined for any mPlane information element, should be used before sending data to a RI component.
The following example, shows how the pinger capability defined in the above probe example be transformed in a standard mPlane message. 

```javascript
/* serialize the capability in an mPlane RI message*/
var RIprobeCapability = probeCapability.to_dict();

/* Shows the serialized capability*/
console.log(RIprobeCapability);
```

This is the output of the above example

```json
{
    "capability": "measure",
    "label": "This_is_a_pinger",
    "metadata": {
        "System_type": "Just a TEST",
        "System_version": "0.1a",
        "System_ID": "Pinger"
    },
    "link": "",
    "token": "ead75d3d198ce8f45ae371246da15fd89d05de24",
    "when": "now ... future / 1s",
    "resultvalues": [],
    "results": [
        "delay.twoway"
    ],
    "parameters": {
        "destination.ip4": "192.168.0.0/16",
        "source.ip4": "192.168.0.1"
    }
}
```


#Transport

The base mPlane library implements the mPlane Information Model elements that can be used to build messages to be carried on top of any protocol of choice (SSH, HTTP,...).
A nodejs HTTPS API has been implemented for basic usage (create and read elements) and can be find [here](https://github.com/finvernizzi/mplane_http_transport.git).

#EXAMPLES

## Supervisor
A complete supervisor implementation written in javascript is available [here](https://github.com/finvernizzi/supervisor).


## Reasoner
A reasoner written for DEMO purpouses is available [here](https://github.com/finvernizzi/reasoner)

## Probe
A probe implementing ping, traceroute and HTTP download delay is available [here](https://github.com/finvernizzi/mplane_probe)

#Documentation
Please refer to the API reference [mPlane nodejs API](http://finvernizzi.github.io/mplane/)

#Release Notes

You can find release notes [here](https://github.com/finvernizzi/mplane/blob/master/release_notes.md)

#LICENSE
This software is released under the [BSD](https://en.wikipedia.org/wiki/BSD_licenses#2-clause_license_.28.22Simplified_BSD_License.22_or_.22FreeBSD_License.22.29) license.


release_notes.md
#0.0.5

- Specification, Statement, Exception from_dict functions now correctly return a new specification from the received JSON
- thrown error better descriptions
- a Parameter can be created from a Parameter (obvious but now also implemented...)
- a Specification can be created from any compatible obj (even from a Specification)
- Element now correctly initializes also from an obj ( _ bug) (with constrains)
- A Parameter is now a direct descendant of an Element. You can also directly create a Parameter
- Value of a Parameter can be undefined. isValid and met_by are not applied in this case
- _mpcv_hash debug. Now correctly reads parameters and constraints. Also metavalues are correctly considered (type and value)
- md5 replaced by sha1 in token generation
- If no name is provided for a parameter, Statemet defines a unique name for it
- Statement.getAllParameters
- Statement.getParameterNames
- Constraints getters
- Constraints.UNDEF now is null. This helps a lot in checking if is defined
- Constraints creator cleaned and debugged
- From a Specification now is possible to get a parameter value (getParameterValue)
- When we create an Element from a registry we memorize also the registry object id in _registryOID. It could be usefull also to register the registry we are using.
- ResultColumn now correctly inherits from Element in nodeJs way
- Statement resultcolumn rewritten and debugged.
- Elements can be directly created by json with the mandatory param type indicating the name of the oid from the registry we want to load.
- Constraints: unparse_constraints is now a global, not prototype, function
- Compliant with the standard messages of the reference implementation (Capability , Specification , Result ). 
    - Sending a message does not require any specifica action
    - Decoding of a received message (Statement) is done by means of mPlane.from_dict, that returns an instance of an object contained in the message.
      This same function can be of use for create any type of object from a JSON description
- Parameter constructor now works also with _value
- cloning of objects instead of simple copy in Specification
- Constraints params now correctly set by addConstraint when the config is a String
- from_dict can receive a string or an object. If the object is a Buffer (the result of a http request) it directly convert to JSON. This is a very common case.
- result and ResultValues correctly adapted in rows and column format of the RI (from_dict and to_dict)
- BUG scripting!

#0.6.0

- Can read TI and standard RI registry format
- Added primitive address
- bool primitive is now named boolean

#0.6.2

- Can read registry over HTTP, HTTPS (or locally) using sync-request. Use file://<full path to registry file> for local registry

#0.7.1

- Access to a registry (file or https, http) in a dedicated function
- Support for registry-uri/includes
- Registry access errror management

#0.7.3
- Now all elements from different registries all all pushed in the array


#0.7.5
- Debug on undefined primitives

#0.7.6
- Exception better supported
- isException for simple check of the returned object
- Exception.prototype.toString()
- Some debug in from_dict

#0.7.7
- [BUG] Capability token in to_dict is correctly set
- Better work in regitry chains using merge

#0.7.8
- [BUG] REAL prim was not provided by all the needed functionalities
- Registry merge simplified. Merge library dependency removed
- Added some messages for checking the registry chain has been loaded and merged


Examples/pinger.js
/**
 * Pinger probe implementing the Capability push, Specification pull model
 *
 * @author fabrizio.invernizzi@telecomitalia.it
 * @version 0.0.1a
 *
 */

var exec = require('child_process').exec,
    mplane = require('mplane'),
    _ = require("lodash"),
    request = require("request"),
    http = require("http"),
    inquirer = require("inquirer"),
    os=require('os'),
    url = require('url'),
    prettyjson = require('prettyjson'),
    async = require("async");

var ifaces=os.networkInterfaces();
var ipAdresses = [];
for (var dev in ifaces) {
    var alias=0;
    ifaces[dev].forEach(function(details){
        if (details.family=='IPv4') {
            //console.log(dev+(alias?':'+alias:''),details.address);
            ipAdresses.push(details.address);
            ++alias;
        }
    });
}

var __MY_IP__ = "127.0.0.1";
var supervisorIP = "127.0.0.1";
var supervisorPort = 2426;
var specificationLink = "";

var questions = [
    {
        type: "list",
        name: "ipSource",
        message: "Select your source IP address",
        choices: ipAdresses,
        filter: function( val ) { return val.toLowerCase(); }
    }
];

prompt = inquirer.prompt( questions, function( answers ) {
    __MY_IP__ = answers.ipSource;

    // Initialize available primitives from the registry
    mplane.Element.initialize_registry("registry.json");

    // Source address and destination parameters initialized from the registry
    var sourceAddress = new mplane.Element("source.ip");
    //  My source address should never be changed!
    sourceAddress.addConstraint( __MY_IP__ );

    var destinationAddress = new mplane.Element("destination");
    destinationAddress.addConstraint("10.34.32.0 ... 10.34.32.255");

    var numberOfEchoRequests = new mplane.Element("number");
    numberOfEchoRequests.addConstraint("1 ... 10");

    // Ping result - MEAN RTT
    var meanRTT = new mplane.Element("delay.twoway");

    var statement = new mplane.Statement({verb:mplane.Statement.VERB_MEASURE});
    statement.add_parameter("My source address" , sourceAddress , __MY_IP__);

    // Here we should set a valid ip address in accordance with the constraints on the destinationAddress
    statement.add_parameter("The destination address" , destinationAddress , __MY_IP__);
    statement.add_parameter("The number of echo requests the pinger should send" , numberOfEchoRequests , 1);
    statement.add_result_column("Mean RTT", meanRTT);

    // THE CAPABILITY we will announce
    var capability = new mplane.Capability(statement);

    registerToSupervisor(capability);
    checkForSpecifications(10000);
});


function mean(values){
    var sum = 0 , elements = 0;
    _.each(values , function(val , index){
        if (!_.isNaN(val)){
            sum += val*1;
            elements +=1;
        }
    });
    return (sum/elements);
}

function doAPing(destination , Wait , requests , callback){
 exec("ping -S " + __MY_IP__ + "  -W "+ Wait  +" -c " + requests + " " + destination  + " | grep from",
  function (error, stdout, stderr) {
      var times = [];
    if (!stdout)
        console.log("No answer")
    else{
        var replies = stdout.split(/\n/);
        _.each(replies , function(row , index){
            process.stdout.write(". ");
            var vals = row.split(/[\t\s]+/);
            _.each(vals, function(el , index){
                var element = el.split("=");
                switch(element[0]){
                    case "time":
                        times.push(element[1]);
                        break;
                    default:
                        // nothing to do here
                }
            });
        });
        console.log(mean(times))
        callback(null, mean(times));
        //console.log("Mean RTT:"+mrtt);
        //return mrtt;
    }
    if (error !== null) {
      callback(error , null);
    }
  });
}

/**
 * POST to the supervisor all the probe information
 * @param capability
 */
function registerToSupervisor( capability) {
    // Serialize the capability Object
    var post_data = capability.to_dict();

    var post_options = {
        path: '/register',
        method: 'POST',
        host: supervisorIP,
        port: supervisorPort,
        followRedirect:false,
        headers: {
            'Content-Type': 'application/json',
            'Content-Length': post_data.length
        }
    };

    // Set up the request
    var post_req = http.request(post_options, function(res) {
        // We expect a redirect to the Specification url
        if (res.statusCode == 302){
            console.log("The supervisor answered with:");
            console.log(res.headers.location);
            specificationLink = res.headers.location;
        }else{
            console.log("Unable to register to the supervisor. May be we are already registered?("+res.statusCode+")");
        }
    });

    // post the data
    post_req.write(post_data);
    post_req.end();
}

function checkForSpecifications(period){
    setInterval(function(){

        specificationUrlDetails = url.parse(specificationLink, true);
        var get_options = {
            path: specificationUrlDetails.path,
            method: 'GET',
            host: specificationUrlDetails.hostname,
            port: specificationUrlDetails.port,
            followRedirect:false
        };
        if (!specificationUrlDetails){
            console.log("ERROR")
            return;
        }

        var req = http.get(get_options, function(res) {
            // Buffer the body entirely for processing as a whole.
            var bodyChunks = [];
            var spec = null;
            res.on('data', function(chunk) {
                bodyChunks.push(chunk);
            }).on('end', function() {
                body = Buffer.concat(bodyChunks);
                if (body == "NOTHING for you"){
                    console.log("NO specification ... nothing to do!");
                }else{
                    spec = JSON.parse(body);
                    var dest = spec._params['The destination address']._value;
                    var reqNum = spec._params['The number of echo requests the pinger should send']._value;

                    async.waterfall([
                        function(callback){
                            console.log("------------------------------------------");
                            console.log("So i should work...")
                            doAPing(dest, 4 , reqNum , callback);
                        }
                    ], function (err, meanRTT) {
                        console.log("CALCULATED:"+meanRTT);
                        postResultToSupervisor( spec , meanRTT);

                    });

                }
            })
        });

        req.on('error', function(e) {
            console.log('ERROR: ' + e.message);
        });

    } , period);
}

function postResultToSupervisor( specification , meanRTT) {
    var specificationUrlDetails = url.parse(specificationLink, true);
    var specSplit = specificationUrlDetails.path.split("=");
    var token = specSplit[1];

    // Create a result obj from the specification
    var result = new mplane.Result(specification);
    // Add the result comlumns
    if (_.isNaN(meanRTT))
        meanRTT = 0;
    result.add_result_column("meanRTT" , new mplane.Element("delay.twoway")).setValue({"meanRTT" : meanRTT});

    var post_data = result.to_dict();
    var post_options = {
        path: '/result/token='+token,
        method: 'POST',
        host: supervisorIP,
        port: supervisorPort,
        followRedirect:false,
        headers: {
            'Content-Type': 'application/json',
            'Content-Length': post_data.length
        }
    };

    // Set up the request
    var post_req = http.request(post_options, function(res) {
        if (res.statusCode == 200){
            console.log("The supervisor received the result of my work!");
        }else{
            console.log("Unable to send result to the supervisor ("+res.statusCode+")");
        }
        // Here since it is async
        console.log("------------------------------------------");
    });

    //console.log(post_data)
    // post the data
    post_req.write(post_data);
    post_req.end();

    return;
}

Examples/supervisor.js
/**
 * mPlane supervisor example
 *
 * @author fabrizio.invernizzi@telecomitalia.it
 * @version 0.0.1a
 *      Capability push, Specification pull paradigm ONLY
 *
 */


var mplane = require('mplane'),
    //cli = require('cli').enable('status','catchall'),
    express = require('express'),
    app = express(),
    connect = require('connect'),
    //prompt = require('prompt'),
    inquirer = require("inquirer"),
    _ = require("lodash"),
    prettyjson = require('prettyjson');
   // router = express.Router();


var LISTEN_PORT = 2426;
var LISTEN_IP = "127.0.0.1";

// List of probes registerd to this supervisor
// Each probe is identified by its unique token
var __registered_probes__ = {};
// List of required measures, indexed by probe token
// Capability push, Specification pull paradigm
var __required_specifications__ = {};


var __SPEC_STATUS_QUEUED_="queued";
var __SPEC_STATUS_PROBE_TAKEN_="taken";
var __SPEC_STATUS_GOT_RESULT="got_result";

// ---------------------------------------------------------------
// WEB server
// ---------------------------------------------------------------
app.use(connect.bodyParser({strict:false}));

/**
 * We have received a POST message with a register
 *
 */
// FIXME: the register path is not in the architecture specification
app.post('/register', function(req, res){
    var newToken = registerProbe(req.body);
    if (newToken){
        res.writeHead(302, {
            'Location': "http://"+LISTEN_IP+":"+LISTEN_PORT+"/specification/token="+newToken
        });
        res.end();
    }
    else{
        res.writeHead( 403, 'Already registered', {'content-type' : 'text/plain'});
        res.end( 'Already registered');
    }
});
var tokenReq = null;
app.param('token', function(req,res, next, token){
    tokenReq = null;
    var t = token.split("=");
    if (t[1])
        tokenReq = t[1];
    next();
});
app.get('/specification/:token', function(req, res){
    var r = "NOTHING for you";
    if (tokenReq) {
        if (__required_specifications__[tokenReq]){
            var specHashes = _.keys(__required_specifications__[tokenReq]);
            for (var specNum=0; specNum<specHashes.length; specNum++){
                var specHash = specHashes[specNum];
                if ((__required_specifications__[tokenReq][specHash].specification_status == __SPEC_STATUS_QUEUED_)) {
                    var spec = createSpecificationFromRegisteredSpecification(tokenReq, specHash);
                    // We need to be sure the spec is using the right token
                    spec.set_token(specHash);
                    __required_specifications__[tokenReq][specHash].specification_status = __SPEC_STATUS_PROBE_TAKEN_;
                    res.json(spec);
                    // One a time, so stop
                    specNum = specHashes.length;
                    r=null;
                }
            }
        }
    }
    if (r)
        res.send(403 , r);
});
app.post('/result/:token', function(req, res){
    var r = "NOTHING for you";
    var result = new mplane.Result(req.body);
    if (tokenReq) {
        __required_specifications__[tokenReq][result.get_token()] = result;
        __required_specifications__[tokenReq][result.get_token()].specification_status = __SPEC_STATUS_GOT_RESULT;
        res.send("OK");
    }
    else{
        res.send(403 , r);
    }
});

var server = app.listen(LISTEN_PORT , LISTEN_IP);


function registerProbe(probe){
    //FIXME: the get_token function is not exposed in mplane
    if (!__registered_probes__[probe._token]){
        __registered_probes__[probe._token] = probe;
        return probe._token;
    }
    return false;
}


// ---------------------------------------------------------------
// Start the prompt
// ---------------------------------------------------------------
motd();
cli();


var prompt;

function cli(){
    var prompt = "mPlane - "+LISTEN_IP+"@"+LISTEN_PORT+"#";
    var questions = [
        {
            type: "list",
            name: "cmd",
            message: prompt,
            choices: [ "Show", "Specifications" , new inquirer.Separator() , "Quit" ],
            filter: function( val ) { return val.toLowerCase(); }
        },
            {
                type: "list",
                name: "showType",
                message: prompt+" (SHOW) ",
                choices: [ "Probes", "Specifications" ],
                when: function(curCmd){ return (curCmd.cmd == "show");},
                filter: function( val ) { return val.toLowerCase(); }
            },
            {
                type: "list",
                name: "probeHash",
                message: prompt+"(SPECIFICATION) " ,
                choices: function(){
                    var h = _.keys(__registered_probes__);
                    if (h.length == 0)
                        return ["-- No Probes yet --"];
                    else
                        return h;
                },
                when: function(curCmd){return (curCmd.cmd == "specifications");},
                filter: function( val ) { return val.toLowerCase(); }
            }

    ];

    prompt = inquirer.prompt( questions, function( answers ) {
        switch (answers.cmd){
            case "quit":
                console.log("\nSEE YOU NEXT TIME!\n\n");
                server.close();
                return false;
                break;
            case "show":
                doShow(answers.showType);
                console.log("\n\n");
                cli();
                break;
            case "specifications":
                if (answers.probeHash == "-- no probes yet --"){
                    cli();
                    break;
                }else{
                    queueSpecification(answers.probeHash);
                    console.log("\n\n");
                }
                //cli();
                break;
            default:
               // cli();
        }
    });
}


function doShow(args){
    switch (args.toLowerCase()){
        case "probes":
            if (_.keys(__registered_probes__).length == 0){
                console.log("---> NO probes registered YET <---");
                return true;
            }
            _.forEach(__registered_probes__, function(vale , key , collection){
                console.log("-----------------------------------------------------------");
                console.log("---> ["+key+"] <---");
                console.log(prettyjson.render(__registered_probes__[key]));
                //console.log(__registered_probes__[key]);
            });
            return true;
            break;
        case "specifications":
            if (_.keys(__required_specifications__).length == 0){
                console.log("---> NO specification queued <---");
                return true;
            }else{
                showSpecifications();
                return true;
            }
            break;
        default:
            return true;
    }

    return true;
}

/**
 * Defines a probe measure
 * The number is internal and is derived from the registration order
 * From the number i can derive the token
 * @param probeNumber
 */
function queueSpecification(probeHash){
    if (typeof  probeHash == "undefined"){
        console.log("You should choose a probe");
        return true;
    }
    if (!__required_specifications__[probeHash])
        __required_specifications__[probeHash] = {}; // We can have multiple request for a single probe

    var spec = createSpecificationFromRegisteredProbe(probeHash);

    var inputParams = [ ];
    var paramNames = spec.parameter_names();

    var parameters = [];
    for (var i=0; i<paramNames.length ; i++) {
       parameters.push(spec.getParameter(paramNames[i])) ;
    }

    var msg = "";
    var defaultValue = "";
    //FIXME: it should work also with multiple contraints!!!
    _.forEach(parameters , function(par){
        var constr = par.getConstraints();
        var pn = par.getName().replace(/\s/gi , "_");

        switch (constr['0']._type){
            case "singleton":
                defaultValue = constr['0']._param;
                break;
            case "range":
                defaultValue = constr['0']._param.valA;
                break;
            case "list":
                defaultValue = constr['0']._param[0];
                break;
            default:
        }
        msg = par.getDescription() +" [" + constr['0'].unParse_constraint()+"]";

        inputParams.push(
            {
                type: "input",
                name: pn,
                message: msg || "TBD",
                validate: function(value){return (constr['0'].met_by(value));},
                default:defaultValue
            }
        )
    });

    // New prompt for input about parameters to be filled
    var paramsCli = inquirer.prompt( inputParams, function( userParams ) {
        //console.log(userParams);//OK!
        // Do not confuse the name of the parameter in the specification parameters obj with the parameter name in the parameter itself!
        for (var parNum=0 ; parNum<paramNames.length ; parNum++) {
            par = parameters[parNum];
            try {
                var parName = par.getName();
                //console.log(parName.replace(/\s/gi , "_"));
                //console.log(userParams[parName.replace(/\s/gi , "_")]);
                par.setValue(userParams[parName.replace(/\s/gi, "_")]);
                spec.setParameterValue(paramNames[parNum], userParams[parName.replace(/\s/gi, "_")]);
                spec.specification_status = __SPEC_STATUS_QUEUED_;
            } catch (e) {
                console.log(e)
            }
        }
        var specToken = spec.update_token();

        __required_specifications__[probeHash][specToken]= {};
        __required_specifications__[probeHash][specToken]= spec;
        console.log("Specification for "+probeHash+" QUEUED ("+specToken+")");
        cli();
    });
}

/**
 * Show queued specifications
 * @param probeNumber
 * @returns {boolean}
 */
function showSpecifications(probeHash){
    if (_.keys(__registered_probes__).length == 0){
        console.log("---> NO probes registered YET <---");
        return true;
    }

    // Are we asking for a specific probe or for all probes?
    if (!_.isUndefined(probeHash)){
        _.forEach(__registered_probes__ , function(hash) {
            showSpecificationForProbe(hash);
        });
    }
    else{
        var hashes = _.keys(__registered_probes__);
        _.forEach(hashes , function(hash){
            if (__required_specifications__[hash]){
                showSpecificationForProbe(hash);
                console.log( "\n");
            }
        });

        return true;
    }
}

function showSpecificationForProbe(probeHash){
    console.log("+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++");
    console.log("Queued specification for "+probeHash);
    var specHashes = _.keys(__required_specifications__[probeHash]);
    for (var specNum=0 ; specNum<specHashes.length ; specNum++){
        var specToken = specHashes[specNum];
        console.log("------> Specification "+specToken+" <-------");
        console.log(prettyjson.render(__required_specifications__[probeHash][specToken]));
        //console.log(prettyjson.render(specification));
    }
    //_.forEach(__required_specifications__[probeHash] , function(specification){
//        console.log("------> Specification "+specification.get_token()+" <-------");
//        console.log(prettyjson.render(specification));
        //console.log(specification);
    //});
    console.log("+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++");
}

function help(){
    console.log("SHOW probes");
    console.log("HELP");
}


function motd(){
    console.log("\n\n");
    console.log("           mPlane supervisor DEMO");
    console.log("\n\n");

}

function createSpecificationFromRegisteredProbe(probeHash){
    var statement = new mplane.Statement({verb:mplane.Statement.VERB_MEASURE});
    var params = __registered_probes__[probeHash]._params;
    var parNames = _.keys(params);
    _.forEach(parNames , function(name){
        var paramNew = new mplane.Element(params[name]._element);
        statement.add_parameter(name , paramNew , params[name]._value);
    });
    return new  mplane.Specification(statement);
}

function createSpecificationFromRegisteredSpecification(probeHash , specHash){
    specification = __required_specifications__[probeHash][specHash];
    var statement = new mplane.Statement({verb:mplane.Statement.VERB_MEASURE});
    var params = specification._params;
    var parNames = _.keys(params);
    _.forEach(parNames , function(name){
        var paramNew = new mplane.Element(params[name]._element);
        statement.add_parameter(name , paramNew , params[name]._value);
    });
    return new  mplane.Specification(statement);
}